packages/std/src/memory.rs

```diff
@@ -69,15 +69,84 @@ pub unsafe fn consume_region(ptr: *mut Region) -> Vec<u8> {
     )
 }
 
+/// Element that can be used to construct a new `Box<Region>`
+///
+/// # Safety
+///
+/// The following invariant must be upheld:
+///
+/// - full allocated capacity == value returned by capacity
+///
+/// This is important to uphold the safety invariant of the `dealloc` method, which requires us to pass the same Layout
+/// into it as was used to allocate a memory region.
+///
+/// And since `size` is one of the parameters, it is important to pass in the exact same capacity.
+///
+/// See: <https://doc.rust-lang.org/stable/alloc/alloc/trait.GlobalAlloc.html#safety-2>
+pub unsafe trait RegionSource {
+    fn ptr(&self) -> *const u8;
+    fn len(&self) -> usize;
+    fn capacity(&self) -> usize;
+}
+
+unsafe impl RegionSource for &[u8] {
+    fn ptr(&self) -> *const u8 {
+        self.as_ptr()
+    }
+
+    fn len(&self) -> usize {
+        (*self).len()
+    }
+
+    fn capacity(&self) -> usize {
+        self.len()
+    }
+}
+
+unsafe impl RegionSource for Vec<u8> {
+    fn ptr(&self) -> *const u8 {
+        self.as_ptr()
+    }
+
+    fn len(&self) -> usize {
+        self.len()
+    }
+
+    fn capacity(&self) -> usize {
+        self.capacity()
+    }
+}
+
+unsafe impl<T: ?Sized> RegionSource for &T
+where
+    T: RegionSource,
+{
+    fn ptr(&self) -> *const u8 {
+        (**self).ptr()
+    }
+
+    fn len(&self) -> usize {
+        (**self).len()
+    }
+
+    fn capacity(&self) -> usize {
+        (**self).capacity()
+    }
+}
+
 /// Returns a box of a Region, which can be sent over a call to extern
 /// note that this DOES NOT take ownership of the data, and we MUST NOT consume_region
 /// the resulting data.
 /// The Box must be dropped (with scope), but not the data
-pub fn build_region(data: &[u8]) -> Box<Region> {
-    let data_ptr = data.as_ptr() as usize;
+pub fn build_region<S>(data: S) -> Box<Region>
+where
+    S: RegionSource,
+{
+    // Well, this technically violates pointer provenance rules.
+    // But there isn't a stable API for it, so that's the best we can do, I guess.
```

Comment:
Which provenance rules are you referring to?

The whole `memory` module is wasm32 only, such that we know
- The memory is a Wasm linear memory where memory can be addressed by offsets
- usize is 32 bit long

Reply:
Yeah, makes sense, but it's just internally a bunch of hints for the compiler and miri.  
It contains stuff like the address size, address, and some other metadata which a `usize` can technically not accurately represent since it's just a pointer somewhere in memory.

It's technically just an experiment but they also wanna explore if it makes sense for the compiler to enforce provenance at some point AFAIK.

<https://doc.rust-lang.org/nightly/std/ptr/index.html#strict-provenance>

But as the docs say

> The following text is non-normative, insufficiently formal, and is an extremely strict interpretation of provenance. It’s ok if your code doesn’t strictly conform to it.

Just something that was on my mind and I commented in ¯\\_(ツ)_/¯  
We can remove the comment, too.

Reply:
Thanks, all good, no worries. I was just not sure if there is anything practically relevant we need to be careful of.